evm_arithmetization/src/prover.rs

```diff
@@ -63,9 +63,11 @@ where
     C: GenericConfig<D, F = F>,
 {
     timed!(timing, "build kernel", Lazy::force(&KERNEL));
+    let inputs2 = inputs.clone();
+    let inputs3 = inputs.clone();
     let (registers_before, registers_after, mut memory_before) =
-        generate_segment::<F>(max_cpu_len, segment_index, &inputs)?;
-    let mut state = GenerationState::<F>::new(inputs.clone(), &KERNEL.code)
+        generate_segment::<F>(max_cpu_len, segment_index, inputs)?;
+    let mut state = GenerationState::<F>::new(inputs2, &KERNEL.code)
```

Comment:
Are we actually shaving off a clone? To me it just looks like we moved it elsewhere.

Reply:
Also calling them `inputs, inputs2, inputs3` is kinda misleading, they are just clones of the initial inputs, maybe just calling the clone inside the function call would be cleaner

Reply:
The one-less-clone is [here](https://github.com/0xPolygonZero/zk_evm/pull/81/files#diff-c4cbd7b12f2ed0e2a1d85ad499c53cb0488be5435d2c227213f75bebda2922beL149) and I think that it would be possible to remove at least one more, but it'd be a deeper refactor.

As for naming and collecting all clones in one place, I guess that's debatable but I think it's less bad to show all clones that are taking place in one spot. I hope that they will be sufficiently annoying to someone (hint hint) to motivate removing them. I'd argue that cloning in the method call makes it harder to see exactly what is being copied over the course of a rather big method.

Reply:
I also think we're just moving around the clones (the one for `input2` is the one saved on `geenrate_segment`). On the other hand, it seems that `GenerationState::<F>::new` should also work if it only takes a reference to the public inputs, if we want to save a clone.